include/simdjson/inline/document.h

```diff
@@ -591,24 +597,38 @@ WARN_UNUSED inline bool document::parser::set_max_depth(size_t max_depth) {
 
   if (!ret_address || !containing_scope_offset) {
     // Could not allocate memory
-    return false;
+    return MEMALLOC;
   }
 
   _max_depth = max_depth;
-  return true;
+  return SUCCESS;
 }
 
-WARN_UNUSED
-inline error_code document::parser::init_parse(size_t len) noexcept {
-  if (len > capacity()) {
-    return error = CAPACITY;
+WARN_UNUSED inline bool document::parser::allocate_capacity(size_t capacity, size_t max_depth) noexcept {
+  return !set_capacity(capacity) && !set_max_depth(max_depth);
+}
+
+inline error_code document::parser::ensure_capacity(size_t desired_capacity) noexcept {
+  // If we don't have enough capacity, (try to) automatically bump it.
+  if (unlikely(desired_capacity > capacity())) {
+    if (desired_capacity > max_capacity()) {
+      return error = CAPACITY;
+    }
+
+    error = set_capacity(desired_capacity);
```

Comment:
I am mildly worried about how we could end up being constantly reallocating... e.g., if someone parses ever growing documents... but cleverer ways would probably backfire.

Reply:
Yes, the worst case is someone continuously parses ever larger documents. It's fairly pathological for servers and config files, but I'm sure it could happen somewhere! If the user knows their case, though, they can set the maximum at 0 and then call set capacity to get current behavior.

In practice, in a server where you keep the parser around and parse many requests with unknown sizes, the problem is just hard. How do you keep exceptionally large documents from permanently chewing up tons of memory? How do I even decide what exceptionally large means? There is stuff we can help with there, I think, but it needs more thought and experiment. The recommendation I have right now is to set a cap on the parser auto allocation and when you get a capacity error, spin up a new temporary parser instead of growing the current one.

There are cleverer ways to auto allocate, like how vectors work ... but vectors aren't designed for the "approach a steady state" use case parsers will typically see, so it's probably not the right model.